### PR TITLE
Add xbox gamepad support for Windows

### PR DESCRIPTION
**Pull request by DragonRhythm**
**ptz-controls.cpp/.hpp**
- Save off a use_gamepad in the profile for the checkbox.
- The instance of the gamepad class is created and stored here.
- Expose camera controls as public for the gamepad to use. 

**ptz-gamepad.cpp/.hpp**
- Gamepad class handles all gamepad input and calls the appropriate ptzcontrols functions with data.
- Thread class is spun up when the gamepad checkbox is marked.
- The thread runs and polls the gamepad using xinput.
- Signals and slots are used to transmit the data back to the same thread as controls.

**ptz-onvif.cpp/.hpp**
- Had to rename a class due to a conflict with xinput. 

**settings.cpp/.hpp**
- Handle the new checkbox, set the new checkbox from the profile. 

**settings.ui**
- Add a new tab in the settings for the gamepad.

Signed-off-by: Eric Schmidt <dragonrhythm@gmail.com>

### CODE REVIEW

**Comment by glikely** [comment]
Hi DragonRhythm. Thanks for looking into this.

I've run the GitHub CI action on the pull request, and as you can see it fails on Linux and MacOS builds. Those will need to be addressed before I can merge the patch. You don't need to enable gamepad support on those platforms, but you do need to make sure those builds are not broken before it can be merged. You can see the build failures in the CI log.

**Review by glikely — CHANGES_REQUESTED** [review]
Hi DragonRhythm. I've taken some time to read through the diff and have some comments below. Thanks again for working on gamepad support

File: README.md
Comment (DragonRhythm) [reply README.md:149]: I was thinking of something similar. At the very least it needs to make sure stop commands get through.

File: src/ptz-controls.cpp
Comment (DragonRhythm) [reply src/ptz-controls.cpp:338]: Understood. Where I work this is standard practice so I thought nothing of it.

File: src/ptz-controls.cpp
Comment (DragonRhythm) [reply src/ptz-controls.cpp:490]: siblingAtRow doesn't wrap

File: src/ptz-gamepad.cpp
Comment (DragonRhythm) [reply src/ptz-gamepad.cpp:341]: I didn't copy QGamepad on purpose to make it my own. However, I will update this to follow a registration api in controls.

File: src/ptz-onvif.hpp
Comment (DragonRhythm) [reply src/ptz-onvif.hpp:75]: ok

File: src/ptz-controls.cpp
Comment (DragonRhythm) [reply src/ptz-controls.cpp:504]: I'd prefer to leave this one so it can wrap.

**Comment by DragonRhythm** [comment]
Ready for another look and run. I couldn't figure out how to run the formatter locally to fix that up though.

**Comment by glikely** [comment]
> Ready for another look and run.

Awesome, thanks! I'm hoping to get some time to review today or tomorrow. From a quick look this is looking better.

One thing that will help is if the commits are reorganised. Instead of committing additional changes on top, please squash the fixups down into a single commit so that it is a single logical change. Otherwise I have to look at multiple commits to work out the entire change.

Second, it makes it easier to review if distinct changes split out into separate commits. For example, if the rename of DeviceCapabilities is unrelated (aside from the conflict) to the rest of the changes. If it is in a separate commit, then it simplifies the gamepad diff, and I can even merge that portion while the other pieces are still under review.

>I couldn't figure out how to run the formatter locally to fix that up though.

I've pushed a commit out to mainline that displays the diff when the format check fails. That should help.

**Comment by DragonRhythm** [comment]
I tried to squash them but it tells me "Unable to squash. Squashing replays all commits up to the last one required for the squash. A merge commit cannot exist among those commits." I had to resolve a merge conflict earlier so maybe that's why?

**Comment by glikely** [comment]
> I tried to squash them but it tells me "Unable to squash. Squashing replays all commits up to the last one required for the squash. A merge commit cannot exist among those commits." I had to resolve a merge conflict earlier so maybe that's why?

I'd avoid merge commits for patches under development. Merging is more appropriate for maintainers pulling in trees from other developers, but for a patch series under review it makes things more complicated. Try a rebase instead:

```
$ git rebase origin/main
... you may need to fixup a few things., after which do:
$ git add <fixed files>
$ git rebase continue
```

Then you'll be free of the merge commits and you can squash or reorganise freely

File: src/ptz-controls.cpp
Comment (glikely) [reply src/ptz-controls.cpp:506]: Fair enough. I hadn't been thinking about wrapping around at the start or end of the list.

File: src/ptz-gamepad.cpp
Comment (glikely) [reply src/ptz-gamepad.cpp]: 😁 I understand wanting to make it your own. However from a maintenance perspective the closer it is to QGamepad the better. It is an established API that is well abstracted.

In fact, my preferred solution would be to import QGamepad and fix the bugs rather than writing something new. That would give a path for enabling support in MacOS and Linux. But, you're doing this work, not me. I'm not going to reject your gamepad support just based on that preference.

Comment (DragonRhythm) [reply src/ptz-gamepad.cpp]: It's much more similar in the newest revision.

Comment (glikely) [reply src/ptz-gamepad.cpp]: 👍

File: src/ptz-onvif.cpp
Comment (glikely) [inline src/ptz-onvif.cpp:472]: Since the ONVIF rename is a self contained change, I've got ahead and applied it as a separate commit.

File: src/settings.cpp
Comment (glikely) [inline src/settings.cpp:270]: `state` isn't unused

**Review by glikely — CHANGES_REQUESTED** [review]

**Comment by glikely** [comment]
Thank you for your work on this. I'm not merging this PR, but it did help a lot in getting my head around how to add joystick support. I decided instead to import the QJoystick code because it works across Windows, MacOS and Linux, and it has already worked out a lot of the corner cases.

Joystick support is now in v0.15.0. Please give it a try and let me know how it works for you.

**Pull request closed without merging**
